RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,268 @@
+* If one of them is `undefined`, it evaluates to `undefined`.
+
+#### ^
+Consider the meta-expression `b ^ e`, where `e` is a rational.
+* If `b` is `well-formed`, it evaluates to a `well-formed` unit resulting from raising `b` to the power `e`.
+* If `b` is `empty`, it evaluates to `empty`.
+* If `b` is `undefined`, it evaluates to `undefined`.
+
+#### der
+Consider the meta-expression `der(e)`.
+* If `e` is `well-formed`, it evaluates to `e * ("s" ^ -1)`.
+* If `e` is `empty`, it evaluates to `empty`.
+* If `e` is `undefined`, it evaluates to `undefined`.
+
+This definition prevents `der` from introducing `"s"` in a model that does not contain any units.
+It could lead to unwanted unit errors otherwise.
+
+### Evaluation
+Unit variables can't be found to be `empty` or `undefined` unit.
+This means that after evaluation, a unit meta-expression is either `undefined`, `empty` or does not contain any `undefined` or `empty` unit.
+
+## Unit compatibility
+Two well-formed units are said to be unit compatible if all of the following are true:
+* Their base unit factorizations are equal.
+* Their base unit offsets are equal.
+* Their base unit scalings are equal.
+
+Examples:
+- `"N"` and `"m/s2"` have different base unit factorizations, they are not compatible.
+- `"K"` and `"degC"` have different base unit offsets, they are not compatible.
+- `"s"` and `"ms"` have different base unit scaling, they are not compatible.
+- `"kN"` and `"W.s/mm"` are compatible.
+
+It is a quality of implementation how equality of offsets and scalings are checked.
+
+`empty` units and `undefined` units are always compatible with each other, themselves and any `well-formed` units.
+
+## Unit convertibility
+A well-formed unit `u1` is said to be convertible to another well-formed unit `u2` if:
+* Their base unit factorizations are equal.
+
+Examples:
+- `"N"` is not convertible to `"m/s2"`, since they have different base unit factorizations.
+- `"K"` is convertible to `"degC"`, since they differ by base unit offsets.
+- `"s"` is convertible to `"ms"`, since they differ by base unit scaling.
+- `"kN"` is convertible to `"W.s/mm"`.
+
+`empty` units and `undefined` units are never convertible to each other, themselves or any `well-formed` units.
+
+## Unit inference
+
+### Modified Hindley-Milner
+
+The original Hindley-Milner algorithm is designed to infer types, and consists of building a sequence of substitutions that returns the type of any expression of a given program.
+Applying this algorithm to perform unit inference in Modelica comes with two important differences:
+  * Not all variables must have a unit in a given model
+  * There are operation on units, that is, not all constraints are of the form  _`var1.unit` is compatible with `var2.unit`_.
+
+The following describes a modified version of the Hindley-Milner's algorithm to perform unit inference in the Modelica language.
+
```

Comment:
This would make a better heading for the item list:
```suggestion
Hindley-Milner's algorithm modified to perform unit inference in the Modelica language:
```